From ccdd7600fa6ea3b85ce1b316d98b0518eedf610c Mon Sep 17 00:00:00 2001
From: Eric Sandeen <sandeen@redhat.com>
Date: Thu, 29 May 2014 21:31:54 -0400
Subject: [fs] xfs: xfs_readsb needs to check for magic numbers

Message-id: <5387A74A.3020107@redhat.com>
Patchwork-id: 81193
O-Subject: [RHEL7.1 PATCH] xfs: xfs_readsb needs to check for magic numbers
Bugzilla: 1102940
RH-Acked-by: Brian Foster <bfoster@redhat.com>
RH-Acked-by: Dave Chinner <dchinner@redhat.com>

This is for:

Bug 1102940 - fs: xfs_readsb needs to check for magic numbers during probe-based mounts (edit)

The following patch is on the xfs list & reviewed, but not yet upstream:

xfs: xfs_readsb needs to check for magic numbers

From: Dave Chinner <dchinner@redhat.com>

Commit daba542 ("xfs: skip verification on initial "guess"
superblock read") dropped the use of a verifier for the initial
superblock read so we can probe the sector size of the filesystem
stored in the superblock. It, however, now fails to validate that
what was read initially is actually an XFS superblock and hence will
fail the sector size check and return ENOSYS.

This causes probe-based mounts to fail because it expects XFS to
return EINVAL when it doesn't recognise the superblock format.

cc: <stable@vger.kernel.org>
Reported-by: Plamen Petrov <plamen.sisi@gmail.com>
Tested-by: Plamen Petrov <plamen.sisi@gmail.com>
Signed-off-by: Dave Chinner <dchinner@redhat.com>
Reviewed-by: Christoph Hellwig <hch@lst.de>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/xfs/xfs_mount.c b/fs/xfs/xfs_mount.c
index 993cb19..a2f4459 100644
--- a/fs/xfs/xfs_mount.c
+++ b/fs/xfs/xfs_mount.c
@@ -323,8 +323,19 @@ reread:
  /*
   * Initialize the mount structure from the superblock.
   */
- xfs_sb_from_disk(&mp->m_sb, XFS_BUF_TO_SBP(bp));
- xfs_sb_quota_from_disk(&mp->m_sb);
+ xfs_sb_from_disk(sbp, XFS_BUF_TO_SBP(bp));
+ xfs_sb_quota_from_disk(sbp);
+
+ /*
+  * If we haven't validated the superblock, do so now before we try
+  * to check the sector size and reread the superblock appropriately.
+  */
+ if (sbp->sb_magicnum != XFS_SB_MAGIC) {
+  if (loud)
+   xfs_warn(mp, "Invalid superblock magic number");
+  error = EINVAL;
+  goto release_buf;
+ }
 
  /*
   * We must be able to do sector-sized and sector-aligned IO.
@@ -337,11 +348,11 @@ reread:
   goto release_buf;
  }
 
- /*
-  * Re-read the superblock so the buffer is correctly sized,
-  * and properly verified.
-  */
  if (buf_ops == NULL) {
+  /*
+   * Re-read the superblock so the buffer is correctly sized,
+   * and properly verified.
+   */
   xfs_buf_relse(bp);
   sector_size = sbp->sb_sectsize;
   buf_ops = loud ? &xfs_sb_buf_ops : &xfs_sb_quiet_buf_ops;
-- 
1.7.1